uhd: build with utils by default

hexagonal-sun opened the pull request:
## Description of changes

Build the uhd package with utils enabled by default. The increase in the
closure size is relatively small, circa 18%. However, this is a very
large package to be manually compiled, especially on older systems.

On balance, it makes sense to build the uhd utilities as the default.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

bjornfor commented:
How much bigger is this new variant? Is big size the reason it's not enabled by default?

hexagonal-sun commented:
@bjornfor  The new variant is 3Mb larger:

With utils:
```
513M	/nix/store/pswa9hvssjg11mjk97a2610ayinanijq-uhd-4.4.0.0
```

Without utils:
```
510M	/nix/store/5k6wsnayv7maj1mdwmhrqhir5sawnynz-uhd-4.4.0.0
```

I added this option as a separate package so that people could save themselves the space/bandwidth if they don't need it.  But given that it's a less than 1% increase in size, I wonder if there is an argument for making the `uhd` package include utils by default?

bjornfor commented:
Note that the closure size difference is bigger:

```
$ du -sch $(nix-store -qR /nix/store/4d0zv8qliz8bwysrhm4f7wwzg4vagxry-uhd-4.4.0.0) | grep total
652M    total
$ du -sch $(nix-store -qR /nix/store/mbyymjcgxdbc85r79n39l92vphfs7pgp-uhd-4.4.0.0) | grep total
771M    total
```

(18% increase.)

But personally I'd rather have one uhd package with enableUtils = true than to split them. Given that uhd itself is ~500 MB, adding an extra package isn't exactly free.

hexagonal-sun commented:
Agreed.  I've modified the PR, proposing the make `enableUtils = true` the default for the `uhd` package.

bjornfor commented:
I edited the PR title, but the PR body is out of sync.

doronbehar reviewed (commented):
Thanks for the PR! I am happy to see people use uhd directly (and not only as gnuradio's dependency), and that these flags are of use :). I am rather sure that if you'd compare `nix-store --query --references $(nix-build -A uhd)`, with and without utils enabled, you'll notice that now Python is there as well. Could you please verify that? I'm asking because I want `gnuradio{,3_{8,9}}Minimal` derivations to not reference python, even if indirectly.

We could take this PR further and create a `uhd` and `uhdMinimal` derivations, with more and less features enabled in each accordingly. Then, we could write something like:

```nix
gnuradioMinimal = callPackage .../wrapper.nix {
  unwrapped = callPackage .../default.nix {
    uhd = uhdMinimal;
  };
};
```

hexagonal-sun commented:
Indeed, it does reference python:

```
matthew@xin ~/D/nixpkgs (add-uhd-with-utils)> nix-store --query --references ./result
/nix/store/905gkx2q1pswixwmi1qfhfl6mik3f22l-glibc-2.37-8
/nix/store/gi26p79iq8jrw51irq5x82c2cqlgicxi-gcc-12.3.0-lib
/nix/store/d834dlcbmagn5h1mggvwm8axmhi6r3x2-python3-3.10.12-env
/nix/store/g0ffqxqwpk5dgjnca16d55v473axb1cx-libusb-1.0.26
/nix/store/ibdxq15h123wv3cyvpqvdidw13vd1b83-boost-1.81.0
/nix/store/pswa9hvssjg11mjk97a2610ayinanijq-uhd-4.4.0.0
```

I could create a new package `uhdMinimal` with utils disabled, then modify the minimal gnuradio components to use that if that would work?

hexagonal-sun commented:
Okay, I've made the appropriate modifications as suggested by @doronbehar .

doronbehar commented on pkgs/applications/radio/uhd/default.nix:
```diff
@@ -13,7 +13,7 @@
-, enableUtils ? false
+, enableUtils ? true
 , libusb1
 , enableDpdk ? false
```
Could we enable this feature by default while at it? Same for `enablePythonApi` - if we include Python references, let's go all the way.

doronbehar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20470,6 +20470,7 @@ with pkgs;
 
   uhd3_5 = callPackage ../applications/radio/uhd/3.5.nix { };
   uhd = callPackage ../applications/radio/uhd { };
+  uhdMinimal = callPackage ../applications/radio/uhd { enableUtils = false; };
```
Could you use `.override` please?

doronbehar reviewed (commented):
Thanks @hexagonal-sun for all the work. If you are having trouble building this on a weak machine, we could use `@ofborg` to build `uhd` and `gnuradioMinimal`.

hexagonal-sun commented:
Thanks for the review!  Those suggestions should now be addressed.

I'm not that familiar with @ofborg.  How can I instruct it to build the changed packages?

doronbehar reviewed (approved):
Diff LGTM!

doronbehar commented:
> Thanks for the review! Those suggestions should now be addressed.
> 

It's my pleasure :) 

> I'm not that familiar with @ofborg. How can I instruct it to build the changed packages?

Oh that's super simple :)

@ofborg build uhd

Since this code doesn't technically change how `gnuradio{,3_{8,9}}Minimal` derivations are evaluated, I didn't ask ofborg to build them. Rather, we are only testing whether `uhd` compiles fine with all of those features enabled, and for all platforms specified in `uhd.meta`.

doronbehar commented:
OK so now I observe compilation errors in ofborg due to dpdk support enabled. It might be possible to fix it via an update, but let's not deal with that here, and add a comment near `enableDpdk ? false` stating that there are compilation errors when this feature is enabled.

hexagonal-sun commented:
> OK so now I observe compilation errors in ofborg due to dpdk support enabled.

Interesting.  Where do you see the status of ofborg builds? I can't seem to see anything in the `Checks` section.

> It might be possible to fix it via an update, but let's not deal with that here, and add a comment near `enableDpdk ? false` stating that there are compilation errors when this feature is enabled.

Agreed.  I've now pushed changes that disabled the Dpdk feature for now.

doronbehar committed:
uhd: Improve comments near disabled features

doronbehar committed:
uhd: build with utils by default

Build the uhd package with utils enabled by default, and disallow it to
retain python references if these features are disabled. Add a new
package uhdMinimal which does not build with these features. Modify the
gnuradio{,3_{8,9}}Minimal packages so that they use the uhd package
without python references.

doronbehar committed:
uhd3_5: remove, unused anymore

doronbehar commented:
@hexagonal-sun the ofborg build logs are available in these links:

![image](https://github.com/NixOS/nixpkgs/assets/10998835/ec8b8df4-4205-4ae2-8a47-da1d5eee23b5)

I pushed here here some more improvements, and I disabled the python API support as it appears broken according to ofborg [link to build log](https://logs.ofborg.org/?key=nixos/nixpkgs.258968&attempt_id=c043f6b5-8b42-498e-993c-b9c4b10e6dbe).

doronbehar commented:
Let's wait a few hours to see if the Darwin builds aren't broken by my changes.

hexagonal-sun commented:
Thanks for merging!

doronbehar commented:
@hexagonal an update to UHD is to 4.5.0.0 is available for your review at #259749 .
